jax_galsim/interpolatedimage.py

```diff
@@ -856,6 +828,73 @@ def _drawKImage(self, image, jac=None):
         # Return an image
         return Image(array=im, bounds=image.bounds, wcs=image.wcs, check_bounds=False)
 
+    @lazy_property
+    def _pos_neg_fluxes(self):
+        # record pos and neg fluxes now too
+        pflux = jnp.sum(jnp.where(self._pad_image.array > 0, self._pad_image.array, 0))
+        nflux = jnp.abs(
+            jnp.sum(jnp.where(self._pad_image.array < 0, self._pad_image.array, 0))
+        )
+        pint = self._x_interpolant.positive_flux
+        nint = self._x_interpolant.negative_flux
+        pint2d = pint * pint + nint * nint
+        nint2d = 2 * pint * nint
+        return [
+            pint2d * pflux + nint2d * nflux,
+            pint2d * nflux + nint2d * pflux,
+        ]
+
+    @property
+    def _positive_flux(self):
+        return self._pos_neg_fluxes[0]
+
+    @property
+    def _negative_flux(self):
+        return self._pos_neg_fluxes[1]
+
+    def _flux_per_photon(self):
+        return self._calculate_flux_per_photon()
+
+    def _shoot(self, photons, rng):
+        # we first draw the index location from the image
+        img = self._pad_image
+        subkey = rng._state.split_one()
+        inds = jrng.choice(
+            subkey,
+            img.array.size,
+            shape=(photons.size(),),
+            replace=True,
+            # we use abs here since some of the pixels could be negative
+            # and for a noise image this procedure results in a fair
+            # sampling of the noise
+            p=jnp.abs(img.array.ravel()) / jnp.sum(jnp.abs(img.array)),
+        ).astype(int)
+        yinds, xinds = jnp.unravel_index(inds, img.array.shape)
+
+        xedges = jnp.arange(img.bounds.xmin, img.bounds.xmax + 2) - 0.5
+        yedges = jnp.arange(img.bounds.ymin, img.bounds.ymax + 2) - 0.5
+
+        # now we draw the position within the pixel
+        ud = UniformDeviate(rng)
+        photons.x = ud.generate(photons.x) + xedges[xinds]
+        photons.y = ud.generate(photons.y) + yedges[yinds]
+        photons.flux = (
+            jnp.sign(img.array.ravel())[inds]
+            * self._flux_per_photon()
+            * (self.positive_flux + self.negative_flux)
+            / photons.size()
+        )
+
+        # accounnt for offset - we add the offset to get to
```

Comment:
This might be wrong. Photons have x,y in pixels and so we may not need to handle the offset here at all?

Reply:
This code is correct. I added a test.